nep-dapi.mediawiki

```diff
@@ -0,0 +1,586 @@
+      break;
+  }
+});
+</pre>
+
+
+
+====getBalance====
+Return balance of a specific asset for the given account.
+
+'''Method Interface'''
+
+<pre>
+function getBalance({ address: string, asset: string }): Promise<{ address: string, asset: string, amount: string }}>
+</pre>
+
+'''Input arguments'''
+
+address: string - The asset symbol the check the balance for the connected account
+
+asset: string - The asset symbol the check the balance for the connected account
+
+
+'''Success return value'''
+
+address: string - Address of the connected account
+
+asset: string - Asset symbol for the requested balance
+
+amount: string - Parsed balance of the requested asset
+
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_DENIED`
+
+'''Example'''
+
+<pre>
+getBalance({
+  address: 'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru',
+  asset: 'NKN'
+})
+.then((result: Result) => {
+  const {
+    address,
+    asset,
+    amount,
+  } = result;
+
+  console.log('Connected address: ' + address);
+  console.log('Asset: ' + asset);
+  console.log('Amount: ' + amount);
+})
+.catch((err: string) => {
+  switch(err) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+====send====
+Invoke a transfer of a specified amount of a given asset from the connected account to another account.
+
+'''Method Interface'''
+
+<pre>
+function send({ to: string, asset: string, amount: string, remark?: string }): Promise<txid: string>
+</pre>
+
+'''Input arguments'''
+
+to: string - Address of the receiver of the assets to be sent
+
+asset: string - The symbol of the asset to be sent
+
+amount: string - The parsed amount of the asset to be sent
+
+remark?: string - (Optional) Description of the transaction to be made
+
+
+'''Success return value'''
+
+txid: string - The transaction ID of the send invocation
+
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_REFUSED`|`SEND_ERROR`|`MALFORMED_ADDRESS`|`CANCELED`
+
+'''Example'''
+
+<pre>
+send({
+  to: 'ATaWxfUAiBcQixNPq6TvKHEHPQum9bx79d',
+  asset: 'GAS',
+  amount: '0.0001',
+  remark: 'Hash puppy clothing purchase. Invoice#abc123',
+})
+.then((txid: string) => {
+  console.log('Send transaction success! Transaction ID: ' + txid);
+})
+.catch((err: string) => {
+  switch(err) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_REFUSED:
+      console.log('Connection dApp not connected. Please call the "connect" function.');
+      break;
+   case SEND_ERROR:
+    console.log('There was an error when broadcasting this transaction to the network.');
+    break;
+   case MALFORMED_ADDRESS:
+      console.log('The receiver address provided is not valid.');
+      break;
+  case CANCELED:
+    console.log('The user has canceled this transaction.');
```

Comment:
How can you cancel a transaction?

Reply:
The implementation for each wallet provider may differ, but there will have to be some mechanism for a user to control what applications the wallet will sign transactions for. It may be the case that the wallet allows it's users to require that every request for signature require a user approval. In this case, the user can review each transfer request made by the dApp via this protocol, and the user may choose to decline a transfer request.